buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionArchiveSetupPlugin.java

```diff
@@ -0,0 +1,196 @@
+ *
+ * This configures the default artifacts for the distribution specific
+ * subprojects. We have subprojects for two reasons:
+ * 1. Gradle project substitutions can only bind to the default
+ *    configuration of a project
+ * 2. The integ-test-zip and zip distributions have the exact same
+ *    filename, so they must be placed in different directories.
+ * 3. We provide a packed and an unpacked variant of the distribution
+ *    - the unpacked variant is used by consumers like test cluster definitions
+ */
+public class InternalDistributionArchiveSetupPlugin implements Plugin<Project> {
+
+    private NamedDomainObjectContainer<DistributionArchive> container;
+
+    @Override
+    public void apply(Project project) {
+        project.getPlugins().apply(BasePlugin.class);
+        registerAndConfigureDistributionArchivesExtension(project);
+        registerEmptyDirectoryTasks(project);
+        configureGeneralTaskDefaults(project);
+        configureTarDefaults(project);
+    }
+
+    private void registerAndConfigureDistributionArchivesExtension(Project project) {
+        container = project.container(DistributionArchive.class, name -> {
+            var subProjectDir = buildTaskToSubprojectName(name);
+            var copyDistributionTaskName = name.substring(0, name.length() - 3);
+            var explodedDist = project.getTasks()
+                .register(copyDistributionTaskName, Copy.class, copy -> copy.into(subProjectDir + "/build/install/"));
+            return name.endsWith("Tar")
+                ? new DistributionArchive(project.getTasks().register(name, SymbolicLinkPreservingTar.class), explodedDist, name)
+                : new DistributionArchive(project.getTasks().register(name, Zip.class), explodedDist, name);
+        });
+        // Each defined distribution archive is linked to a subproject.
+        // A distribution archive definition not matching a sub project will result in build failure.
+        container.whenObjectAdded(distributionArchive -> {
+            var subProjectName = buildTaskToSubprojectName(distributionArchive.getArchiveTask().getName());
+            project.project(subProjectName, sub -> {
+                sub.getPlugins().apply("base");
+                sub.getArtifacts().add("default", distributionArchive.getArchiveTask());
+                var explodedArchiveTask = distributionArchive.getExplodedArchiveTask();
+                var defaultConfiguration = sub.getConfigurations().getByName("default");
+                var publications = defaultConfiguration.getOutgoing();
+                var variant = (ConfigurationVariantInternal) publications.getVariants().maybeCreate("directory");
+                variant.getAttributes().attribute(ArtifactAttributes.ARTIFACT_FORMAT, ArtifactTypeDefinition.DIRECTORY_TYPE);
+                variant.artifactsProvider(() -> Collections.singletonList(new DirectoryPublishArtifact(explodedArchiveTask)));
```

Comment:
Is there no public API to accomplish this?

Reply:
Unfortunately there isn't as far as I can tell. I also reached out to the gradle folks in slack and in the forum at https://discuss.gradle.org/t/registering-a-directory-type-variant-backed-by-a-custom-copy-task/37486/2 but so far no feedback

Reply:
I changed the implementation here to rely on a different configuration for the explodedDist. This allows us to not rely on internal api and makes the setup a bit simpler